### PR TITLE
Api service and database added for tab now playing

### PR DESCRIPTION
**Pull request by julian-mainardis-globant**
## Description

Added the Api service and Database structure. In this PR for the tab "Now Playing"

## Story Type

- New feature :white_check_mark:
- Change request
- Bug fix
- Tech

## How Can This Be Tested?

Playing the emulator

## Evidences

| Pixel 3  | Nexus S |
| ------------- | ------------- |
| pixel3 | nexuss |

## Coverage
image
image
image

## Checklist

- All the unit test passed :white_check_mark:

## Tested On
- Pixel 3
- Nexus S

## Package Structure
image image

### CODE REVIEW

File: app/src/main/java/com/example/moviesapp/MainActivity.kt [hunk 1]
```diff
@@ -1,11 +1,41 @@
...
 import android.os.Bundle
+import androidx.appcompat.app.AppCompatActivity
+import com.example.domain.entity.Movie
+import com.example.moviesapp.databinding.ActivityMainBinding
+import org.koin.androidx.viewmodel.ext.android.viewModel
 
 class MainActivity : AppCompatActivity() {
+
+    private lateinit var binding: ActivityMainBinding
+    private val viewModel by viewModel<MainViewModel>()
```
Comment (mariano98montero) [inline app/src/main/java/com/example/moviesapp/MainActivity.kt]: this should be "by inject()"

File: app/src/main/java/com/example/moviesapp/MainActivity.kt [hunk 2]
```diff
@@ -1,11 +1,41 @@
 package com.example.moviesapp
 
-import androidx.appcompat.app.AppCompatActivity
 import android.os.Bundle
+import androidx.appcompat.app.AppCompatActivity
+import com.example.domain.entity.Movie
+import com.example.moviesapp.databinding.ActivityMainBinding
+import org.koin.androidx.viewmodel.ext.android.viewModel
 
 class MainActivity : AppCompatActivity() {
```
Comment (mariano98montero) [inline app/src/main/java/com/example/moviesapp/MainActivity.kt]: this class should also implement KoinComponent

File: app/src/main/java/com/example/moviesapp/MainActivity.kt [hunk 3]
```diff
@@ -1,11 +1,41 @@
...
+
+    private lateinit var binding: ActivityMainBinding
+    private val viewModel by viewModel<MainViewModel>()
+
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
-        setContentView(R.layout.activity_main)
+        binding = ActivityMainBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+        viewModel.getMovies()
```
Comment (mariano98montero) [inline app/src/main/java/com/example/moviesapp/MainActivity.kt]: this should be below the next line

File: app/src/main/java/com/example/moviesapp/MainViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import com.example.domain.util.Result
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+
+class MainViewModel(private val getNowPlayingMoviesUseCase: GetNowPlayingMoviesUseCase) : ViewModel() {
+
+    private var _movieState = MutableLiveData<MovieData>()
+    val movieState: LiveData<MovieData>
+        get() = _movieState
```
Comment (mariano98montero) [inline app/src/main/java/com/example/moviesapp/MainViewModel.kt]: please remove the _ on variable naming

**Review by mariano98montero — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/moviesapp/MainActivity.kt (hunk 1 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/moviesapp/MainActivity.kt]: done

File: app/src/main/java/com/example/moviesapp/MainActivity.kt (hunk 2 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/moviesapp/MainActivity.kt]: done

File: app/src/main/java/com/example/moviesapp/MainActivity.kt (hunk 3 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/moviesapp/MainActivity.kt]: done

File: app/src/main/java/com/example/moviesapp/MainViewModel.kt (hunk 4 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/moviesapp/MainViewModel.kt]: done

File: app/src/main/java/com/example/moviesapp/activity/MainActivity.kt [hunk 5]
```diff
@@ -21,39 +19,29 @@ class MainActivity : AppCompatActivity(), KoinComponent {
     private val viewModel: MainActivityViewModel by inject()
     private lateinit var binding: ActivityMainBinding
     private lateinit var pageAdapter: PageAdapter
```
Comment (SartiMau) [inline app/src/main/java/com/example/moviesapp/activity/MainActivity.kt]: this variable is not necessary as global as you only use it in the initialize method

File: app/src/main/java/com/example/moviesapp/activity/MainActivity.kt [hunk 6]
```diff
@@ -21,39 +19,29 @@ class MainActivity : AppCompatActivity(), KoinComponent {
...
-    private var isConnected = false
+    private val icons = arrayOf(R.drawable.ic_now_playing, R.drawable.ic_top_rated, R.drawable.ic_upcoming)
```
Comment (SartiMau) [inline app/src/main/java/com/example/moviesapp/activity/MainActivity.kt]: same here

File: app/src/main/java/com/example/moviesapp/fragment/NowPlayingFragment.kt [hunk 7]
```diff
@@ -20,6 +34,31 @@ class NowPlayingFragment : Fragment() {
...
+            NowPlayingViewModel.MovieState.RESPONSE_SUCCESS -> showMovies(movieData.data)
+            NowPlayingViewModel.MovieState.RESPONSE_ERROR -> showError()
+            NowPlayingViewModel.MovieState.RESPONSE_LOADING -> showLoading()
+        }
+    }
+
+    private fun showMovies(movieList: List<Movie>) {
+        binding.nowPlayingLoader.visibility = View.GONE
+        binding.imgMain.visibility = View.GONE
+        binding.nowPlayingRecyclerView.layoutManager = GridLayoutManager(context, 3)
```
Comment (SartiMau) [inline app/src/main/java/com/example/moviesapp/fragment/NowPlayingFragment.kt]: do not hardcode please, you can create a constant in the companion object

File: app/src/main/java/com/example/moviesapp/fragment/NowPlayingFragment.kt [hunk 8]
```diff
@@ -5,11 +5,25 @@ import android.view.LayoutInflater
...
+import org.koin.android.ext.android.inject
 
 class NowPlayingFragment : Fragment() {
 
     private lateinit var binding: FragmentNowPlayingBinding
+    private val viewModel: NowPlayingViewModel by inject()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        viewModel.getMovieState().observe({ lifecycle }, ::updateUI)
```
Comment (SartiMau) [inline app/src/main/java/com/example/moviesapp/fragment/NowPlayingFragment.kt]: these two lines should be called in the OnViewCreated method not in the onCreate

File: app/src/main/res/values/dimens.xml [hunk 9]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <dimen name="zero_dp">0dp</dimen>
```
Comment (SartiMau) [inline app/src/main/res/values/dimens.xml]: are you still using this dimen?

File: app/src/main/res/values/dimens.xml [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    <dimen name="cardview_image_size">150dp</dimen>
+    <dimen name="cardview_image_border">25dp</dimen>
+    <dimen name="cardview_corner_radius">75dp</dimen>
+    <dimen name="cardview_stroke_size">8dp</dimen>
+    <dimen name="cardview_vertical_padding">12dp</dimen>
+    <dimen name="cardview_horizontal_padding">22dp</dimen>
+    <dimen name="cardview_image_width">200dp</dimen>
+    <dimen name="cardview_image_height">170dp</dimen>
+    <dimen name="image_corner_size">30%</dimen>
```
Comment (SartiMau) [inline app/src/main/res/values/dimens.xml]: is this working?

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/moviesapp/activity/MainActivity.kt (hunk 6 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/moviesapp/activity/MainActivity.kt]: done

File: app/src/main/java/com/example/moviesapp/fragment/NowPlayingFragment.kt (hunk 7 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/moviesapp/fragment/NowPlayingFragment.kt]: done

File: app/src/main/res/values/dimens.xml (hunk 9 above)
Comment (julian-mainardis-globant) [reply app/src/main/res/values/dimens.xml]: no, fixed

File: app/src/main/res/values/dimens.xml (hunk 10 above)
Comment (julian-mainardis-globant) [reply app/src/main/res/values/dimens.xml]: not really, removed

File: app/src/main/java/com/example/moviesapp/activity/MainActivity.kt (hunk 5 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/moviesapp/activity/MainActivity.kt]: done

File: app/src/main/java/com/example/moviesapp/fragment/NowPlayingFragment.kt (hunk 8 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/moviesapp/fragment/NowPlayingFragment.kt]: done

File: app/src/main/java/com/example/moviesapp/activity/MainActivity.kt [hunk 11]
```diff
@@ -20,40 +18,28 @@ class MainActivity : AppCompatActivity(), KoinComponent {
...
-        binding.viewPager.adapter = PageAdapter(this)
-        pageAdapter = binding.viewPager.adapter as PageAdapter
-        TabLayoutMediator(binding.tabLayout, binding.viewPager) { tab, position ->
-            tab.text = pageAdapter.getPageTitle(position)
-        }.attach()
-
-        viewModel.getMovies(isConnected)
-        viewModel.getMovieState().observe({ lifecycle }, ::updateUI)
+        viewModel.getMainActivityState().observe({ lifecycle }, ::updateUI)
+        viewModel.getState()
```
Comment (SartiMau) [inline app/src/main/java/com/example/moviesapp/activity/MainActivity.kt]: the name of this method should be initialize or something like that.. a get should response something and here is no response

File: app/src/main/java/com/example/moviesapp/util/Constants.kt [hunk 12]
```diff
@@ -1,11 +1,13 @@
...
 object Constants {
     const val TAB_NOW_PLAYING = "NOW PLAYING"
     const val TAB_TOP_RATED = "TOP RATED"
     const val TAB_UPCOMING = "UPCOMING"
-    const val EMPTY_DATABASE = "EMPTY_DATABASE"
+    const val IMAGE_RADIUS = 20
     const val FIRST_POSITION = 0
     const val SECOND_POSITION = 1
     const val NUMBER_OF_PAGES = 3
+    const val ERROR = "Something got wrong, please try again later"
```
Comment (SartiMau) [inline app/src/main/java/com/example/moviesapp/util/Constants.kt]: if you will show this method on the screen, this should be in the string res, nor here

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/moviesapp/activity/MainActivity.kt (hunk 11 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/moviesapp/activity/MainActivity.kt]: it's true, fixed

File: app/src/main/java/com/example/moviesapp/util/Constants.kt (hunk 12 above)
Comment (julian-mainardis-globant) [reply app/src/main/java/com/example/moviesapp/util/Constants.kt]: done

**Review by Georgina-Fernandez — APPROVED** [review]

**Review by SartiMau — APPROVED** [review]

**Review by mariano98montero — APPROVED** [review]

**Pull request merged**